### PR TITLE
make privateAccess work with generic types and generic instantiations; fix a SIGSEGV

### PR DESCRIPTION
**Pull request by timotheecour**

### CODE REVIEW

File: compiler/semmagic.nim [hunk 1]
```diff
@@ -575,7 +575,8 @@ proc magicsAfterOverloadResolution(c: PContext, n: PNode,
     result = n
   of mPrivateAccess:
     var t = n[1].typ[0]
-    if t.kind in {tyRef, tyPtr}: t = t[0]
+    if t.kind in {tyRef, tyPtr, tyGenericInst}: t = t[0]
```
Comment (Araq) [inline compiler/semmagic.nim]: Still not correct you need something like:

```nim

var indirections = 0
while true:
  case t.kind 
  of tyRef, tyPtr:
    inc indirections 
    t = t[0]
  of tyGenericInst:
    t = t[0] # really? not  t.lastSon ??
  else:
    break

if indirections <= 1:
  ...
```

If the purpose was to skip a single pointer, that is.

Comment (timotheecour) [reply compiler/semmagic.nim]: thanks, should be correct with latest commit; your version still wasn't quite correct but I've corrected it (also `t.lastSon` wouldn't work for the tyGenericInst case as can be seen by looking at `astalgo.debug`)

the new tests i added reflect that

File: compiler/semmagic.nim [hunk 2]
```diff
@@ -575,7 +575,27 @@ proc magicsAfterOverloadResolution(c: PContext, n: PNode,
...
+    var foundPtrLike = false
+    while true:
+      case t.kind
+      of tyRef, tyPtr:
+        if foundPtrLike: break
+        foundPtrLike = true
+        t = t[0]
+      of tyGenericBody: t = t.lastSon
+      of tyGenericInst: t = t[0]
+      of tyGenericInvocation: t = t[0]
```
Comment (Araq) [inline compiler/semmagic.nim]: `tyAlias` is missing I think.

And while you're at it, move all this logic to a `semPrivateAccess` proc.

File: compiler/semmagic.nim [hunk 3]
```diff
@@ -575,7 +575,27 @@ proc magicsAfterOverloadResolution(c: PContext, n: PNode,
...
+      A2[T] = ref object f1: int
+      A3 = ref object f2: int
+      A4 = object f3: int
+    ]#
+    var foundPtrLike = false
+    while true:
+      case t.kind
+      of tyRef, tyPtr:
+        if foundPtrLike: break
+        foundPtrLike = true
```
Comment (Araq) [inline compiler/semmagic.nim]: Btw I think it should follow all pointers much like we deref them all for `objRefRefRef.field`

Comment (timotheecour) [reply compiler/semmagic.nim]: done; 
I didn't know derefs are more than 1 level, is that documented? (=> just filed https://github.com/nim-lang/Nim/issues/18298)

File: compiler/semmagic.nim (hunk 2 above)
Comment (timotheecour) [reply compiler/semmagic.nim]: done (and added test)

**Comment by timotheecour** [comment]
PTAL

File: compiler/ast.nim [hunk 4]
```diff
@@ -1887,6 +1887,26 @@ proc toObject*(typ: PType): PType =
...
+    A1[T] = ref B[T]
+    A2[T] = ref object f1: int
+    A3 = ref object f2: int
+    A4 = object f3: int
+  ]#
+  result = typ
+  while true:
+    case result.kind
+    of tyGenericBody: result = result.lastSon
+    of {tyRef, tyPtr, tyGenericInst, tyGenericInvocation, tyAlias}: result = result[0]
```
Comment (Araq) [inline compiler/ast.nim]:
```suggestion
    of tyRef, tyPtr, tyGenericInst, tyGenericInvocation, tyAlias: result = result[0]
```

**Pull request merged**